### PR TITLE
Add test_versioning_concurrent_multi_object_delete

### PR DESCRIPTION
**Pull request by cfsnyder**
Tests that concurrent multi-object delete requests which specify
the same versioned object instances return successful object
responses within response body.

relates to: https://tracker.ceph.com/issues/56646

Signed-off-by: Cory Snyder <csnyder@iland.com>

### CODE REVIEW

**Comment by cfsnyder** [comment]
I've verified that this fails on a ceph vstart cluster without my patch and passes on a vstart cluster with my patch.

File: s3tests_boto3/functional/test_s3.py [hunk 1]
```diff
@@ -2014,6 +2014,40 @@ def _make_objs_dict(key_names):
...
+    versions = client.list_object_versions(Bucket=bucket_name)['Versions']
+    eq(len(versions), num_objects)
+    objs_dict = {'Objects': [dict((k, v[k]) for k in ["Key", "VersionId"]) for v in versions]}
+
+    def do_request(n):
+        response = client.delete_objects(Bucket=bucket_name, Delete=objs_dict)
+        assert 'Errors' not in response
+        eq(len(response['Deleted']), num_objects)
+        response = client.list_objects(Bucket=bucket_name)
+        assert 'Contents' not in response
```
Comment (cbodley) [inline s3tests_boto3/functional/test_s3.py:2042]: it doesn't look like these assertions actually cause the test to fail. i changed this function to always assert:
```python
    def do_request(n):
        assert False
```
but it still passes consistently:
```
~/s3-tests $ S3TEST_CONF=s3tests.conf.SAMPLE virtualenv/bin/nosetests -x -v s3tests_boto3/functional/test_s3.py:test_versioning_concurrent_multi_object_delete

s3tests_boto3.functional.test_s3.test_versioning_concurrent_multi_object_delete ... Exception in thread Thread-1:
Traceback (most recent call last):
  File "/usr/lib64/python3.6/threading.py", line 916, in _bootstrap_inner
    self.run()
  File "/usr/lib64/python3.6/threading.py", line 864, in run
    self._target(*self._args, **self._kwargs)
  File "/home/cbodley/s3-tests/s3tests_boto3/functional/test_s3.py", line 2039, in do_request
    assert False
AssertionError

Exception in thread Thread-4:
Traceback (most recent call last):
  File "/usr/lib64/python3.6/threading.py", line 916, in _bootstrap_inner
    self.run()
  File "/usr/lib64/python3.6/threading.py", line 864, in run
    self._target(*self._args, **self._kwargs)
  File "/home/cbodley/s3-tests/s3tests_boto3/functional/test_s3.py", line 2039, in do_request
    assert False
AssertionError

Exception in thread Thread-5:
Traceback (most recent call last):
  File "/usr/lib64/python3.6/threading.py", line 916, in _bootstrap_inner
    self.run()
  File "/usr/lib64/python3.6/threading.py", line 864, in run
    self._target(*self._args, **self._kwargs)
  File "/home/cbodley/s3-tests/s3tests_boto3/functional/test_s3.py", line 2039, in do_request
    assert False
AssertionError

Exception in thread Thread-2:
Traceback (most recent call last):
  File "/usr/lib64/python3.6/threading.py", line 916, in _bootstrap_inner
    self.run()
  File "/usr/lib64/python3.6/threading.py", line 864, in run
    self._target(*self._args, **self._kwargs)
  File "/home/cbodley/s3-tests/s3tests_boto3/functional/test_s3.py", line 2039, in do_request
    assert False
AssertionError

Exception in thread Thread-3:
Traceback (most recent call last):
  File "/usr/lib64/python3.6/threading.py", line 916, in _bootstrap_inner
    self.run()
  File "/usr/lib64/python3.6/threading.py", line 864, in run
    self._target(*self._args, **self._kwargs)
  File "/home/cbodley/s3-tests/s3tests_boto3/functional/test_s3.py", line 2039, in do_request
    assert False
AssertionError

ok

----------------------------------------------------------------------
Ran 1 test in 0.404s

OK
```

**Comment by cfsnyder** [comment]
any reason why this never got merged @cbodley ?

**Comment by ormandj** [comment]
@cbodley ping on this one, we'd like to see this merged

**Review by cbodley — APPROVED** [review]

**Comment by cbodley** [comment]
cherry-picked to ceph-master (fyi @alimaredia)

**Pull request merged**